### ISSUE TITLE
Non-deterministic parallel reductions in Engine tensor ops break reproducibility

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#131]
## Summary

Identical invocations of `Engine.TensorMatMul`, `Engine.TensorSumOfSquares`, and related reduction operations produce different floating-point results across runs on the same inputs, even with identical tensor data and identical upstream seeds. This appears to be caused by parallel SIMD reductions where the order of partial accumulations is not stable between runs — floating-point addition is non-associative, so any variation in accumulation order produces a slightly different final value.

## Impact

In the AiDotNet HRE research work (PR in the main AiDotNet repo), a training test that runs `HRELanguageModel` forward/backward on Tiny Shakespeare with a **seeded** data shuffle produces final validation losses that drift between runs:

- Run 1: final val loss 4.3483
- Run 2: final val loss 7.0079

Same seed, same model init, same shuffle order, same data. The only source of non-determinism is the Engine reductions inside the forward pass.

This forced us to weaken the test's "learned below uniform floor" assertion from `FinalValLoss < log(V)` to `MinTrainLoss < log(V)` — i.e., we can only assert that the model dropped below the floor *at some point* during training, not that it finished below it. That's a real expressiveness loss for the paper's headline claim.

## Repro steps

In a downstream project using AiDotNet.Tensors 0.33.0 + AiDotNet master (feature/harmonic-resonance-engine-clean):

```csharp
// Run the Shakespeare integration test twice with identical config
dotnet test --filter "FullyQualifiedName~SpectralTargetPropagation_TinyShakespeare_ReducesValLoss"
dotnet test --filter "FullyQualifiedName~SpectralTargetPropagation_TinyShakespeare_ReducesValLoss"
```

Observe that the reported `Final val loss` differs between runs despite:
- `HRELanguageModel` constructed with fixed `seed: 42`
- `trainLoader.GetBatches(..., seed: config.Seed + epoch)` — seeded shuffle
- No other source of randomness in the training loop

The divergence is concentrated in late-training steps where accumulated floating-point drift has compounded.

## Affected ops (observed so far)

- `Engine.TensorMatMul` — used in attention-output and target-projection hot paths
- `Engine.TensorSumOfSquares` — used for L2 norm and VerifyFinite checks
- Any other reduction that parallelizes across tiles/threads

Not yet confirmed but suspected: `TensorSum`, `TensorSoftmaxRows` (the softmax normalizer is a reduction), `TensorLogSoftmax`.

## Requested fix

One of:

1. **Deterministic mode flag.** Add `AiDotNetEngine.DeterministicMode` (or equivalent) that forces reductions to run in a fixed order — single-threaded, or multi-threaded with stable tree reduction. Opt-in so the performance hit is only paid when needed (tests, scientific reproducibility).

2. **Documented guarantee that a specific reduction order is stable.** If the current parallelism is deterministic for a fixed hardware config (same core count, same SIMD width), document that explicitly so callers know what to expect.

3. **Fixed-seed reduction tiling.** Allow callers to pass a reduction strategy / tile size so results are reproducible given a fixed strategy, even if not between hardware configs.

Option 1 is the most useful for research reproducibility — the HRE training tests need it, and any downstream research project using the Engine for iterative training will eventually hit the same issue.

## Workaround in place

We changed the HRE test assertion from `FinalValLoss < floor` to `MinTrainLoss < floor` so it's only sensitive to *some* point during training reaching the target, not the final state. Works but weakens the research claim.

---

Filing this at the request of the maintainer as part of the HRE research work. Happy to provide more detailed repro, test traces, or a minimal isolated test case if useful.
